Add check for non-date modified values when harvesting

janette opened the pull request:
connects #2612

## QA Steps
- [ ] Create a harvest source using https://open.fda.gov/data.json
- [ ]  Confirm that the modified values are saved and do not cause errors (view the error log)
- [ ] View the harvest 'datasets' tab `admin/dkan/harvest/dashboard/datasets`, verify the changes make sense

## Change to Harvested Datasets view
Removed the filter for modified 'Date', added a few others
![harvest_view](https://user-images.githubusercontent.com/314172/44877904-0b8bf580-ac6b-11e8-8fad-e6fd929d5f28.png)


## Reminders

- [ ] There is test for the issue.
- [ ] Coding standards checked.
- [ ] Review docs.getdkan.com (or in /docs) to see if it still covers the scope of the PR and update if needed.

janette commented:
If I can get access to the S3 box I can update the test .json to include a term modified value instead of a date value

dafeder commented on modules/dkan/dkan_harvest/dkan_harvest.module:
```diff
@@ -1460,3 +1460,8 @@ function dkan_harvest_delete_confirmation_submit($form, $form_state) {
   dkan_harvest_process_source_datasets($selected_sources, $datasets_op);
 }
+
+function validateDate($date, $format = 'Y-m-d') {
```
CamelCase functions don't follow Drupal naming convention. Should be dkan_harvest_validate_date()

dafeder replied:
CamelCase functions don't follow Drupal naming convention. Should be dkan_harvest_validate_date(). Also, I think this function should live in dkan_dataset, since it's used by that module as well.

dafeder commented on modules/dkan/dkan_dataset/dkan_dataset.module:
```diff
@@ -1304,7 +1304,21 @@ function dkan_dataset_modified_date($node) {
-    $date = format_date($node->field_harvest_source_modified->value(), 'iso_8601_date');
+    // Check if the value is a valid date or term.
+    $d = validateDate($node->field_harvest_source_modified->value());
+    $terms = array_keys(dkan_dataset_content_types_iso_frecuency_map());
```
This isn't sufficient to validate ISO 8601 strings as specified for the modified field in the POD schema. 

> Dates should be ISO 8601 of highest resolution. In other words, as much of `YYYY-MM-DDThh:mm:ss.sTZD` as is relevant to this dataset. If there is a need to reflect that the dataset is continually updated, ISO 8601 formatting can account for this with repeating intervals. For instance, `R/P1D` for daily, `R/P2W` for every two weeks, and `R/PT5M` for every five minutes.

The limited values we provide for Frequency wouldn't support all of the possibilities there.

We should be able to parse it by checking if it starts with `R/` and then grabbing the string after that and running it through PHP's `DateInterval::format()`. See http://us1.php.net/manual/en/class.dateinterval.php

janette replied:
@dafeder also this sounds like we would just display the 'R/PT5M' value on the dataset and not 'Every 5 minutes' then?

janette commented:
Can't find what is triggering this

![error](https://user-images.githubusercontent.com/314172/44877401-7b00e580-ac69-11e8-8d4c-ccf733c4344a.png)

dafeder commented:
1. @janette can you paste in the JSON you're using that produces that error?
2. The decimal problem seems to be a [longstanding bug in PHP](https://bugs.php.net/bug.php?id=53831). We could get around it by using [this library](https://github.com/snroki/SnrokiDateIntervalBundle). I'm going to set this aside for today and let's discuss who can continue with it and when, as this probably adds an additional 2+ hours to this task.

janette commented:
test json: https://api.myjson.com/bins/9lr0k


```
{
@context: "https://project-open-data.cio.gov/v1.1/schema/catalog.jsonld",
@id: "http://demo.getdkan.com/data.json",
@type: "dcat:Catalog",
conformsTo: "https://project-open-data.cio.gov/v1.1/schema",
describedBy: "https://project-open-data.cio.gov/v1.1/schema/catalog.json",
dataset: [
{
@type: "dcat:Dataset",
accessLevel: "public",
contactPoint: {
fn: "Stefanie Gray",
hasEmail: "mailto:stefanie@example.com"
},
description: "harvest modified date as a term",
distribution: [
{
@type: "dcat:Distribution",
downloadURL: "https://s3.amazonaws.com/dkan-default-content-files/phpunit/GenChart_0_0.csv",
mediaType: "text/csv",
format: "csv",
description: "<p>Population of state employees broken down into Gen X, Gen Y, Baby Boomers and Over 65+.</p> ",
title: "TEST - Workforce By Generation (2011-2015)"
}
],
identifier: "95f8eac4-fd1f-4b35-8472-5c87e9425dfa",
keyword: [
"demographics",
"socioeconomic",
"workforce"
],
license: "http://opendefinition.org/licenses/gfdl/",
issued: "2016-06-22",
modified: "R/P1D",
temporal: "2011/2015",
publisher: {
@type: "org:Organization",
name: "TEST - State Economic Council"
},
accrualPeriodicity: "irregular",
title: "TEST - dataset 1",
dataQuality: true
},
{
@type: "dcat:Dataset",
accessLevel: "public",
contactPoint: {
fn: "Stefanie Gray",
hasEmail: "mailto:stefanie@example.com"
},
description: "Modified value with a period",
distribution: [
{
@type: "dcat:Distribution",
downloadURL: "https://s3.amazonaws.com/dkan-default-content-files/phpunit/2015EligibleVsActual.csv",
mediaType: "text/csv",
format: "csv",
description: "<p>This data, ranging from fiscal year 2011 to fiscal year 2015, depicts the percentage of the workforce that is eligible to retire compared to the percentage of the workforce who actually retired.</p> ",
title: "TEST - Retirements: Eligible vs. Actual"
}
],
identifier: "11f8eac4-fd1f-4b35-8472-5c87e9425dfa",
keyword: [
"demographics",
"socioeconomic",
"workforce"
],
accrualPeriodicity: "R/P1M",
license: "http://opendefinition.org/licenses/gfdl/",
modified: "R/P0.5M",
temporal: "2011/2015",
publisher: {
@type: "org:Organization",
name: "TEST - State Economic Council"
},
title: "TEST - dataset 2",
dataQuality: true
},
{
@type: "dcat:Dataset",
accessLevel: "public",
contactPoint: {
fn: "Stefanie Gray",
hasEmail: "mailto:stefanie@example.com"
},
description: "harvest modified date is date, should save as is",
distribution: [
{
@type: "dcat:Distribution",
downloadURL: "https://s3.amazonaws.com/dkan-default-content-files/phpunit/2015EligibleVsActual.csv",
mediaType: "text/csv",
format: "csv",
description: "test"
}
],
identifier: "12f8eac4-fd1f-4b35-8472-5c87e9425dfa",
license: "http://opendefinition.org/licenses/gfdl/",
modified: "2016-08-22",
temporal: "2011/2015",
accrualPeriodicity: "R/P1W",
publisher: {
@type: "org:Organization",
name: "TEST - date modified value"
},
title: "TEST - dataset 3",
dataQuality: true
}
]
}
```

dafeder commented:
(@janette edited your last comment to make this page less cumbersome -- the URL should be all I need thanks)

janette commented:
### Errors on frequency validation

![errors___dkan](https://user-images.githubusercontent.com/314172/45320299-d94e8380-b507-11e8-9b41-5a4f4387646b.png)

dafeder commented:
> I am wondering about the 'Release Date', when there is no issued date and the modified value is an iso code, then the release date will be the date harvested, that seems wrong.

That's a good point -- I thought `issued` was required but it is not. Have we been seeing harvests with no issued dates? If so, we would maybe want to hide this field in that case.

dafeder commented:
@janette did you last commit fix the field_frequency validation issue? Looks like it to me, but want to make sure I'm not missing anything before merging.

janette commented:
@dafeder yes, the field_frequency validation issue was addressed, ready for merge
